File: examples/vulnerable_code.py
"""
Example file with various security vulnerabilities for testing.
This file intentionally contains security issues for demonstration purposes.
"""

import hashlib
import pickle
import subprocess
import sqlite3


# SQL Injection vulnerabilities
def get_user_unsafe(user_id):
    """Vulnerable to SQL injection via string formatting."""
    query = f"SELECT * FROM users WHERE id = {user_id}"
    return database.execute(query)


def search_products_unsafe(search_term):
    """Vulnerable to SQL injection via string concatenation."""
    query = "SELECT * FROM products WHERE name LIKE '%" + search_term + "%'"
    return database.execute(query)


# Cross-Site Scripting (XSS) vulnerabilities  
def render_comment_unsafe(comment):
    """Vulnerable to XSS via unsafe template rendering."""
    from flask import render_template_string
    template = f"<div class='comment'>{comment}</div>"
    return render_template_string(template)


# Weak cryptography
def hash_password_weak(password):
    """Uses weak MD5 hashing algorithm."""
    return hashlib.md5(password.encode()).hexdigest()


def generate_token_weak():
    """Uses weak random number generation."""
    import random
    return str(random.randint(100000, 999999))


# Hardcoded secrets
DATABASE_PASSWORD = "super_secret_password123"
API_KEY = "ak_1234567890abcdef1234567890abcdef"
JWT_SECRET = "my-jwt-secret-key"


def connect_to_database():
    """Contains hardcoded database credentials."""
    connection_string = f"postgresql://user:{DATABASE_PASSWORD}@localhost/mydb"
    return connection_string


# Insecure deserialization
def load_user_data_unsafe(data):
    """Vulnerable to code execution via pickle deserialization."""
    return pickle.loads(data)


# Command injection
def backup_file_unsafe(filename):
    """Vulnerable to command injection."""
    command = f"cp {filename} /backup/"
    subprocess.run(command, shell=True)


# Path traversal
def read_user_file_unsafe(filename):
    """Vulnerable to path traversal attacks."""
    with open(f"/uploads/{filename}", 'r') as f:
        return f.read()


# Unsafe eval usage
def calculate_expression_unsafe(expression):
    """Dangerous use of eval() function."""
    return eval(expression)


# LDAP injection
def authenticate_user_unsafe(username, password):
    """Vulnerable to LDAP injection."""
    import ldap
    filter_str = f"(&(uid={username})(password={password}))"
    # This would be used in LDAP search
    return filter_str


# Safe examples for comparison
def get_user_safe(user_id):
    """Safe parameterized query."""
    query = "SELECT * FROM users WHERE id = ?"
    return database.execute(query, (user_id,))


def hash_password_safe(password):
    """Uses secure SHA-256 hashing."""
    import hashlib
    return hashlib.sha256(password.encode()).hexdigest()


def generate_token_safe():
    """Uses cryptographically secure random generation."""
    import secrets
    return secrets.token_urlsafe(32)


if __name__ == "__main__":
    # Test vulnerable functions
    print("This file contains intentional vulnerabilities for testing purposes.")
    print("Do not use these patterns in production code!")

File: examples/vulnerable_code.js
/**
 * Example JavaScript file with various security vulnerabilities for testing.
 * This file intentionally contains security issues for demonstration purposes.
 */

// Cross-Site Scripting (XSS) vulnerabilities
function updateContent(userInput) {
    // Vulnerable: Direct assignment to innerHTML
    document.getElementById('content').innerHTML = userInput;
}

function displayMessage(message) {
    // Vulnerable: Using eval() with user input
    eval('alert("' + message + '")');
}

function renderTemplate(data) {
    // Vulnerable: Template injection
    const template = `<div>${data.content}</div>`;
    document.body.innerHTML = template;
}

// SQL Injection vulnerabilities (in Node.js context)
function getUserUnsafe(userId) {
    // Vulnerable: String concatenation in SQL query
    const query = `SELECT * FROM users WHERE id = ${userId}`;
    return database.query(query);
}

function searchProductsUnsafe(searchTerm) {
    // Vulnerable: Template literal injection
    const query = `SELECT * FROM products WHERE name LIKE '%${searchTerm}%'`;
    return database.query(query);
}

// Weak cryptography
function generateSessionId() {
    // Vulnerable: Using Math.random() for security-sensitive operations
    return Math.random().toString(36).substring(2);
}

function hashPasswordWeak(password) {
    // Vulnerable: Weak hashing (if using MD5 library)
    return md5(password);
}

// Hardcoded secrets
const API_KEY = "ak_1234567890abcdef1234567890abcdef";
const DATABASE_URL = "mysql://user:password123@localhost/mydb";
const JWT_SECRET = "my-super-secret-jwt-key";

function connectToAPI() {
    // Vulnerable: Hardcoded API credentials
    return fetch('https://api.example.com/data', {
        headers: {
            'Authorization': `Bearer ${API_KEY}`
        }
    });
}

// Insecure deserialization
function loadUserData(jsonData) {
    // Vulnerable: Using eval() to parse JSON
    return eval('(' + jsonData + ')');
}

// Command injection (Node.js)
function backupFile(filename) {
    // Vulnerable: Command injection via exec()
    const { exec } = require('child_process');
    exec(`cp ${filename} /backup/`);
}

// Path traversal
function readUserFile(filename) {
    // Vulnerable: No path validation
    const fs = require('fs');
    return fs.readFileSync(`/uploads/${filename}`, 'utf8');
}

// Prototype pollution
function mergeObjects(target, source) {
    // Vulnerable: No prototype pollution protection
    for (let key in source) {
        target[key] = source[key];
    }
    return target;
}

// Regular expression denial of service (ReDoS)
function validateEmail(email) {
    // Vulnerable: Catastrophic backtracking regex
    const emailRegex = /^([a-zA-Z0-9_\.\-])+\@(([a-zA-Z0-9\-])+\.)+([a-zA-Z0-9]{2,4})+$/;
    return emailRegex.test(email);
}

// Unsafe redirect
function redirectUser(url) {
    // Vulnerable: Open redirect
    window.location.href = url;
}

// Safe examples for comparison
function updateContentSafe(userInput) {
    // Safe: Using textContent instead of innerHTML
    document.getElementById('content').textContent = userInput;
}

function getUserSafe(userId) {
    // Safe: Parameterized query (example with placeholder)
    const query = 'SELECT * FROM users WHERE id = ?';
    return database.query(query, [userId]);
}

function generateSessionIdSafe() {
    // Safe: Using crypto.getRandomValues()
    const array = new Uint32Array(1);
    crypto.getRandomValues(array);
    return array[0].toString(36);
}

function loadUserDataSafe(jsonData) {
    // Safe: Using JSON.parse()
    return JSON.parse(jsonData);
}

// Export for testing
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        updateContent,
        displayMessage,
        getUserUnsafe,
        generateSessionId,
        loadUserData,
        // Safe versions
        updateContentSafe,
        getUserSafe,
        generateSessionIdSafe,
        loadUserDataSafe
    };
}